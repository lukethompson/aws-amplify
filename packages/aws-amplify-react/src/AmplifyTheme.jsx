/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import { Hub, ClientDevice } from 'aws-amplify';

export const Container = {
    fontFamily: `-apple-system,
                BlinkMacSystemFont,
                "Segoe UI",
                Roboto,
                "Helvetica Neue",
                Arial,
                sans-serif,
                "Apple Color Emoji",
                "Segoe UI Emoji",
                "Segoe UI Symbol"`,
    textAlign: 'center'
};

export const Section = {
    fontWeight: '200'
};

export const FormSection = {
    display: 'inline-block',
    fontWeight: '200'
};

export const ErrorSection = {
    background: 'orange',
    padding: '0.5em',
    margin: '0.2em auto',
    fontWeight: '300'
};

export const NavBar = {
    padding: '0.5em 0.5em 0.2em',
    marginBottom: '0.5em',
    borderBottom: '1px solid #00fff6'
};

export const NavRight = {
    textAlign: 'right',
    fontSize: '0.8em'
};

export const NavButton = {
    padding: '0',
    border: 'none',
    borderBottom: '1px solid #007bff',
    margin: 'auto 0.3em',
    background: '#fff',
    color: '#007bff'
}

export const SectionHeader = {
    background: '#007bff',
    color: '#fff',
    padding: '0.6em',
    fontSize: '1.2em',
    fontWeight: 'normal',
    textAlign: 'center'
};

export const SectionBody = {
    background: '#fff',
    color: '#000',
    padding: '1em',
    fontSize: '1em'
};

export const SectionFooter = {
    background: '#fff',
    color: '#000',
    padding: '1em',
    fontSize: '0.8em',
    textAlign: 'left',
    borderBottom: '1px solid #00fff6'
};

export const FormRow = {
    marginBottom: '0.5em',
    textAlign: 'left'
};

export const ActionRow = {
    marginTop: '1.5em',
    textAlign: 'right'
};

export const Input = {
    fontSize: '1em',
    fontWeight: '200',
    width: '15em',
    padding: '4px 2px',
    border: '1px solid #aaa'
};

export const Button = {
    fontSize: '1em',
    padding: '0.3em 0.5em',
    margin: 'auto 0.3em',
    border: '1px solid #007bff',
    background: '#fff',
    color: '#007bff'
}

export const Album = {
    display: 'block',
    width: '100%',
    textAlign: 'left'
}

export const AlbumPhoto = {
    display: 'inline-block',
    width: '32.5%',
    margin: '0.2%',
    verticalAlign: 'top'
}

export const Photo = {
    display: 'inline-block',
    width: '32.5%',
    margin: '0.2%',
    verticalAlign: 'top'
}

export const PhotoImg = {
    maxWidth: '100%'
}

export const AlbumText = {
    display: 'inline-block',
    width: '32.5%',
    margin: '0.2%',
    verticalAlign: 'top',
    maxHeight: 320,
    overflow: 'auto'
}

export const Text = {
    display: 'inline-block',
    width: '32.5%',
    margin: '0.2%',
    verticalAlign: 'top',
    maxHeight: 320,
    overflow: 'auto'
}

export const Picker = {
    width: '100%',
    textAlign: 'center'
}

export const PickerPicker = {
    width: '100%',
    textAlign: 'center'
}

export const PickerButton = {
    background: '#fff',
    border: '1px solid #ccc',
    fontWeight: '200',
    width: '100%'
}

export const PickerInput = {
}

export const A = {
    textDecoration: 'underline'
}

export const Label = {
    margin: 'auto 0.5em'
}

export const Pre = {
    textAlign: 'left',
    margin: '0',
    padding: '1em',
    background: '#eee',
    overflow: 'auto'
}

export const Col2 = {
    display: 'inline-block',
    width: '16.6%'
}

export const Col3 = {
    display: 'inline-block',
    width: '24.5%',
    margin: '0.2%'
}

export const Col4 = {
    display: 'inline-block',
    width: '32%',
    margin: '0.2%'
}

export const Col6 = {
    display: 'inline-block',
    width: '49.5%',
    margin: '0.2%'
}

export const Col8 = {
    display: 'inline-block',
    width: '66%',
    margin: '0.2%'
}

export const Col9 = {
    display: 'inline-block',
    width: '74.5%',
    margin: '0.2%'
}

export const Col10 = {
    display: 'inline-block',
    width: '83.6%',
    margin: '0.2%'
}

export const Col12 = {
    display: 'inline-block',
    width: '99.2%',
    margin: '0.2%'
}

export const HalfHeight = {
    height: 320,
    overflow: 'auto'
}

export const Center = {
    textAlign: 'center'
}

export const Hidden = {
    display: 'none'
}

const AmplifyTheme = {
    container: Container,
    navBar: NavBar,
    navRight: NavRight,
    navButton: NavButton,

    section: Section,
    sectionHeader: SectionHeader,
    sectionBody: SectionBody,
    sectionFooter: SectionFooter,

    formSection: FormSection,
    formRow: FormRow,
    actionRow: ActionRow,
    input: Input,
    button: Button,

    errorSection: ErrorSection,

    photo: Photo,
    photoImg: PhotoImg,
    text: Text,
    album: Album,
    albumPhoto: AlbumPhoto,
    albumText: AlbumText,
    picker: Picker,
    pickerPicker: PickerPicker,
    pickerButton: PickerButton,
    pickerInput: PickerInput,

    a: A,
    label: Label,
    pre: Pre,

    col3: Col3,
    col4: Col4,
    col6: Col6,
    col8: Col8,
    col9: Col9,
    col10: Col10,
    col12: Col12,
    halfHeight: HalfHeight,

    center: Center,
    hidden: Hidden
}

class MediaQuery {
    query() {
        const dim = ClientDevice.dimension();
        const { width, height } = dim;

        if (width < 576) {
            AmplifyTheme.albumPhoto = Object.assign({}, AlbumPhoto, Col6);
            AmplifyTheme.photo = Object.assign({}, Photo, Col12);
            AmplifyTheme.albumText = Object.assign({}, AlbumText, Col12);
        } else if (width < 768) {
            AmplifyTheme.albumPhoto = Object.assign({}, AlbumPhoto, Col4);
            AmplifyTheme.photo = Object.assign({}, Photo, Col6);
            AmplifyTheme.albumText = Object.assign({}, AlbumText, Col8);
        } else if (width < 992) {
            AmplifyTheme.albumPhoto = Object.assign({}, AlbumPhoto, Col3);
            AmplifyTheme.photo = Object.assign({}, Photo, Col6);
            AmplifyTheme.albumText = Object.assign({}, AlbumText, Col6);
        } else {
            AmplifyTheme.albumPhoto = Object.assign({}, AlbumPhoto, Col2);
            AmplifyTheme.photo = Object.assign({}, Photo, Col4);
            AmplifyTheme.albumText = Object.assign({}, AlbumText, Col4);
        }

        AmplifyTheme.halfHeight = Object.assign({}, HalfHeight, { height: height / 2 });
        AmplifyTheme.text = Object.assign({}, AmplifyTheme.text, { maxHeight: height / 2 });
        AmplifyTheme.albumText = Object.assign({}, AmplifyTheme.albumText, { maxHeight: height / 3 });
    }

    onHubCapsule() {
        this.query();
    }
}

const mediaQuery = new MediaQuery();
mediaQuery.query();
Hub.listen('window', mediaQuery, 'AmplifyTheme');

export default AmplifyTheme;
