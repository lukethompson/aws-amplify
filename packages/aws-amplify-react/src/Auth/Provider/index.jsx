import React, { Component } from 'react';

import withGoogle from './withGoogle';
import withFacebook from './withFacebook';

export { default as withGoogle, GoogleButton } from './withGoogle';
export { default as withFacebook, FacebookButton } from './withFacebook';

export function withFederated(Comp) {
    const Federated = withGoogle(withFacebook(Comp));

    return class extends Component {
        render() {
            const federated = this.props.federated || {};
            return (
                <Federated {...this.props} {...federated} />
            )
        }
    }
}
