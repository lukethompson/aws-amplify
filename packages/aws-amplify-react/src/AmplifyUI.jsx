/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import React from 'react';

import { JS } from 'aws-amplify';

import AmplifyTheme from './AmplifyTheme';

export const FormSection = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-form-section" style={theme.formSection}>
            {props.children}
        </div>
    )
}

export const SectionHeader = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-section-header" style={theme.sectionHeader}>
            {props.children}
        </div>
    )
}

export const SectionFooter = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-section-footer" style={theme.sectionFooter}>
            {props.children}
        </div>
    )
}

export const SectionBody = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-section-body" style={theme.sectionBody}>
            {props.children}
        </div>
    )
}

export const ActionRow = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-action-row" style={theme.actionRow}>
            {props.children}
        </div>
    )
}

export const FormRow = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-form-row" style={theme.formRow}>
            {props.children}
        </div>
    )
}

export const InputRow = (props) => {
    const theme = props.theme || AmplifyTheme;
    const p = JS.objectLessAttributes(props, 'theme');
    return (
        <FormRow theme={theme}>
            {beforeAfter(
                <input
                    {...p}
                    className="amplify-input"
                    style={theme.input}
                />
            )}
        </FormRow>
    )
}

export const RadioRow = (props) => {
    const id = props.id || '_' + props.value;
    const theme = props.theme || AmplifyTheme;
    return (
        <FormRow theme={theme}>
            <Radio
                {...props}
                id={id}
            />
            <Label
                htmlFor={id}
                theme={theme}
            >{props.placeholder}</Label>
        </FormRow>
    )
}

export const Radio = (props) => {
    const theme = props.theme || AmplifyTheme;
    const p = JS.objectLessAttributes(props, 'theme');
    return beforeAfter(
            <input
                {...p}
                type="radio"
                className="amplify-radio"
                style={theme.radio}
            />
    )
}

export const CheckboxRow = (props) => {
    const id = props.id || '_' + props.name;
    const theme = props.theme || AmplifyTheme;
    return (
        <FormRow theme={theme}>
            <Checkbox
                {...props}
                id={id}
            />
            <Label
                htmlFor={id}
                theme={theme}
            >{props.placeholder}</Label>
        </FormRow>
    )
}

export const Checkbox = (props) => {
    const theme = props.theme || AmplifyTheme;
    const p = JS.objectLessAttributes(props, 'theme');
    return beforeAfter(
        <input
            {...p}
            type="checkbox"
            className="amplify-checkbox"
            style={theme.checkbox}
        />
    )
}

export const MessageRow = (props) => {
    const theme = props.theme || AmplifyTheme;
    return (
        <FormRow theme={theme}>
            <MessageContent theme={theme}>{props.children}</MessageContent>
        </FormRow>
    )
}

export const MessageContent = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <span className="amplify-message-content" style={theme.messageContent}>
            {props.children}
        </span>
    )
}

export const ButtonRow = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-action-row" style={theme.actionRow}>
            <Button {...props} />
        </div>
    )
}

export const Button = (props) => {
    const theme = props.theme || AmplifyTheme;
    const p = JS.objectLessAttributes(props, 'theme');

    return beforeAfter(
        <button {...p} className="amplify-button" style={theme.button}>
            <ButtonContent theme={theme}>{props.children}</ButtonContent>
        </button>
    )
}

export const ButtonContent = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <span className="amplify-button-content" style={theme.buttonContent}>
            {props.children}
        </span>
    )
}

export const NavButton = (props) => {
    const theme = props.theme || AmplifyTheme;
    const p = JS.objectLessAttributes(props, 'theme');

    return beforeAfter(
        <button {...p} className="amplify-nav-button" style={theme.navButton}>
            {beforeAfter(
                <span style={theme.navButtonContent}>{props.children}</span>
            )}
        </button>
    )
}

export const Link = (props) => {
    const theme = props.theme || AmplifyTheme;
    const p = JS.objectLessAttributes(props, 'theme');
    return beforeAfter(
        <a {...p} className="amplify-a" style={theme.a}>{props.children}</a>
    )
}

export const Label = (props) => {
    const theme = props.theme || AmplifyTheme;
    const p = JS.objectLessAttributes(props, 'theme');
    return beforeAfter(
        <label {...p} className="amplify-label" style={theme.label}>{props.children}</label>
    )
}

export const NavBar = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-nav-bar" style={theme.navBar}>
            {props.children}
        </div>
    )
}

export const NavRight = (props) => {
    const theme = props.theme || AmplifyTheme;
    return beforeAfter(
        <div className="amplify-nav-right" style={theme.navRight}>
            {props.children}
        </div>
    )
}

export const beforeAfter = (el) => {
    const style = el.props.style || {};
    const { before, after } = style;
    if (!before && !after) { return el; }

    return (
        <span>
            {before? <span style={before}>{before.content}</span> : null}
            {el}
            {after? <span style={after}>{after.content}</span> : null}
        </span>
    )
}

export const transparent1X1 = 'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==';

export const white1X1 = 'data:image/gif;base64,R0lGODlhAQABAIAAAP7//wAAACH5BAAAAAAALAAAAAABAAEAAAICRAEAOw==';
