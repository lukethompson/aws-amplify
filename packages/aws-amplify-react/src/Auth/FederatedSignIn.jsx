import React, { Component } from 'react';

import { Logger, JS } from 'aws-amplify';
import AmplifyTheme from '../AmplifyTheme';
import {
    FormSection,
    SectionBody
} from '../AmplifyUI';
import {
    GoogleButton,
    FacebookButton
} from './Provider';

const logger = new Logger('FederatedSignIn');

export class FederatedButtons extends Component {
    google(google_client_id) {
        if (!google_client_id) { return null; }

        const { onStateChange } = this.props;
        return <GoogleButton
                google_client_id={google_client_id}
                onStateChange={onStateChange}
              />
    }

    facebook(facebook_app_id) {
        if (!facebook_app_id) { return null; }

        const { onStateChange } = this.props;
        return <FacebookButton
                facebook_app_id={facebook_app_id}
                onStateChange={onStateChange}
                />
    }

    render() {
        const { authState } = this.props;
        if (!['signIn', 'signedOut', 'signedUp'].includes(authState)) { return null; }

        const federated = this.props.federated || {};
        if (JS.isEmpty(federated)) { return null; }

        const { google_client_id, facebook_app_id } = federated;

        const theme = this.props.theme || AmplifyTheme;
        return (
            <div className="amplify-action-row" style={theme.actionRow}>
                {this.google(google_client_id)}
                {this.facebook(facebook_app_id)}
            </div>
        )
    }
}

export default class FederatedSignIn extends Component {
    render() {
        const { federated, authState, onStateChange } = this.props;
        if (!federated) {
            logger.debug('federated prop is empty. show nothing');
            logger.debug('federated={google_client_id: , facebook_app_id: }');
            return null;
        }

        const theme = this.props.theme || AmplifyTheme;
        return (
            <FormSection theme={theme}>
                <SectionBody theme={theme}>
                    <FederatedButtons
                        federated={federated}
                        theme={theme}
                        authState={authState}
                        onStateChange={onStateChange}
                    />
                </SectionBody>
            </FormSection>
        )
    }
}
