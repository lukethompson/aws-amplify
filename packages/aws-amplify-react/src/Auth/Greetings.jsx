/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import React, { Component } from 'react';
import { Auth, I18n, Logger, Hub } from 'aws-amplify';

import AuthPiece from './AuthPiece';
import AmplifyTheme from '../AmplifyTheme';

const logger = new Logger('Greetings');

export default class Greetings extends AuthPiece {
    constructor(props) {
        super(props);

        this.signOut = this.signOut.bind(this);
        this.checkUser = this.checkUser.bind(this);
        this.onHubCapsule = this.onHubCapsule.bind(this);

        this.state = {
            authState: props.authState,
            authData: props.authData
        }

        Hub.listen('auth', this);
    }

    componentDidMount() {
        this._isMounted = true;
        this.checkUser();
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    signOut() {
        Auth.signOut()
            .then(() => this.changeState('signedOut'))
            .catch(err => { logger.error(err); this.error(err); });
    }

    checkUser() {
        const that = this;
        const { authState } = this.state;
        return Auth.currentAuthenticatedUser()
            .then(user => {
                if (!that._isMounted) { return; }
                if (authState !== 'signedIn') {
                    this.setState({
                        authState: 'signedIn',
                        authData: user
                    });
                    this.changeState('signedIn', user);
                }
            })
            .catch(err => {
                if (!that._isMounted) { return; }
                if (!authState || authState === 'signedIn') {
                    this.setState({ authState: 'signIn' });
                    this.changeState('signIn');
                }
            });
    }

    onHubCapsule(capsule) {
        const { channel, payload, source } = capsule;
        if (channel === 'auth') { this.checkUser(); }
    }

    inGreeting(name) { return 'Hello ' + name; }
    outGreeting() { return 'Please Sign In / Sign Up'; }

    userGreetings(theme) {
        const user = this.state.authData;
        const greeting = this.props.inGreeting || this.inGreeting;
        const name = user.name || user.username;
        const message = (typeof greeting === 'function')? greeting(name) : greeting;
        return (
            <div className="amplify-nav-right" style={theme.navRight}>
                <span>{message}</span>
                <button
                    className="amplify-nav-button"
                    style={theme.navButton}
                    onClick={this.signOut}
                >{I18n.get('Sign Out')}</button>
            </div>
        )
    }

    noUserGreetings(theme) {
        const greeting = this.props.outGreeting || this.outGreeting;
        const message = (typeof greeting === 'function')? greeting() : greeting;
        return <div className="amplify-nav-right" style={theme.navRight}>{message}</div>
    }

    render() {
        const { hide } = this.props;
        const { authState } = this.state;
        const signedIn = (authState === 'signedIn');
        const theme = this.props.theme || AmplifyTheme;

        if (hide && hide.includes(Greetings)) { return null; }

        return (
            <div className="amplify-nav-bar" style={theme.navBar}>
                {signedIn? this.userGreetings(theme) : this.noUserGreetings(theme)}
            </div>
        )
    }
}
