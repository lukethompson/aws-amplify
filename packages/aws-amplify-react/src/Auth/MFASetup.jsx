/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import React, { Component } from 'react';
import { Auth, I18n, Logger } from 'aws-amplify';

import AuthPiece from './AuthPiece';
import AmplifyTheme from '../AmplifyTheme';
import {
    FormSection,
    SectionHeader,
    SectionBody,
    SectionFooter,
    InputRow,
    ButtonRow,
    Link
} from '../AmplifyUI';

import QRCode from 'qrcode.react';

const logger = new Logger('MFASetup');

export default class MFASetup extends AuthPiece {
    constructor(props) {
        super(props);

        this._validAuthStates = ['mfaSetup'];
        this.setup = this.setup.bind(this);
        this.showSecretCode = this.showSecretCode.bind(this);
        this.verifyTotpToken= this.verifyTotpToken.bind(this);

        this.state = {
            code: null
        }
    }

    setup() {
        const user = this.props.authData;
        Auth.setupMFA(user).then((data) => {
            logger.debug('secret key', data);
            const code = "otpauth://totp/AWSCognito:"+ user.username + "?secret=" + data + "&issuer=AWSCognito";
            this.setState({code});

        }).catch((err) => logger.debug('mfa setup failed', err));
    }

    verifyTotpToken() {
        const user = this.props.authData;
        const { totpCode } = this.inputs;
        Auth.verifyTotpToken(user, totpCode)
            .then(() => this.changeState('signedIn', user))
            .catch(err => this.error(err));
    }

    showSecretCode(code) {
        if (!code) return null;
        return (
            <div>
                <QRCode value={code}/>
            </div>
        )
    }

    showComponent(theme) {
        const { hide } = this.props;
        if (hide && hide.includes(MFASetup)) { return null; }
        let code = this.state.code;

        return (
            <FormSection theme={theme}>
                <SectionHeader theme={theme}>{I18n.get('MFA Setup')}</SectionHeader>
                <SectionBody theme={theme}>
                    {this.showSecretCode(code)}
                    <ButtonRow theme={theme} onClick={this.setup}>
                        {I18n.get('Get secret key')}
                    </ButtonRow>
                    <InputRow
                        autoFocus
                        placeholder={I18n.get('totp verification token')}
                        theme={theme}
                        key="totpCode"
                        name="totpCode"
                        onChange={this.handleInputChange}
                    />
                    <ButtonRow theme={theme} onClick={this.verifyTotpToken}>
                        {I18n.get('Verify')}
                    </ButtonRow>
                </SectionBody>
                <SectionFooter theme={theme}>
                    <Link theme={theme} onClick={() => this.changeState('signIn')}>
                        {I18n.get('Back to Sign In')}
                    </Link>
                </SectionFooter>
            </FormSection>
        )
    }
}
