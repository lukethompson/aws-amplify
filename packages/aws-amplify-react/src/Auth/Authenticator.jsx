/*
 * Copyright 2017-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import * as React from 'react';
import { Component } from 'react';
import Amplify, { I18n, ConsoleLogger as Logger, Hub } from '@aws-amplify/core';
import Auth from '@aws-amplify/auth';
import Greetings from './Greetings';
import SignIn from './SignIn';
import ConfirmSignIn from './ConfirmSignIn';
import RequireNewPassword from './RequireNewPassword';
import SignUp from './SignUp';
import ConfirmSignUp from './ConfirmSignUp';
import VerifyContact from './VerifyContact';
import ForgotPassword from './ForgotPassword';
import TOTPSetup from './TOTPSetup';

import AmplifyTheme from '../Amplify-UI/Amplify-UI-Theme';
import AmplifyMessageMap from '../AmplifyMessageMap';

import { Container, Toast } from '../Amplify-UI/Amplify-UI-Components-React';

const logger = new Logger('Authenticator');
const AUTHENTICATOR_AUTHSTATE = 'amplify-authenticator-authState';

export default class Authenticator extends Component {
    constructor(props) {
        super(props);

        this.handleStateChange = this.handleStateChange.bind(this);
        this.handleAuthEvent = this.handleAuthEvent.bind(this);
        this.onHubCapsule = this.onHubCapsule.bind(this);

        this._initialAuthState = this.props.authState || 'signIn';
        this.state = { auth: 'loading' };
        Hub.listen('auth', this);
    }

    componentDidMount() {
        const config = this.props.amplifyConfig;
        if (config) {
            Amplify.configure(config);
        }
        this._isMounted = true;
        this.checkUser();
    }

    componentWillUnmount() {
        this._isMounted = false;
    }
    checkUser() {
        const { auth } = this.state;
        if (!Auth || typeof Auth.currentAuthenticatedUser !== 'function') {
            throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
        }
        return Auth.currentAuthenticatedUser()
            .then(user => {
                if (!this._isMounted) { return; }
                if (auth !== 'signedIn') {
                    this.setState({
                        authState: 'signedIn',
                        authData: user
                    });
                    this.handleStateChange('signedIn', user);
                }
            })
            .catch(err => {
                if (!this._isMounted) { return; }
                let cachedAuthState = null;
                try {
                    cachedAuthState = localStorage.getItem(AUTHENTICATOR_AUTHSTATE);
                } catch (e) {
                    logger.debug('Failed to get the auth state from local storage', e);
                }
                const promise = cachedAuthState === 'signedIn'? Auth.signOut() : Promise.resolve();
                promise.then(() => this.handleStateChange(this._initialAuthState))
                    .catch((e) => {
                        logger.debug('Failed to sign out', e);
                    });
            });
    }

    onHubCapsule(capsule) {
        const { channel, payload, source } = capsule;
        if (channel === 'auth' && (payload.event === 'configured' || payload.event === 'cognitoHostedUI')) { 
            this.checkUser(); 
        }
    }

    handleStateChange(state, data) {
        logger.debug('authenticator state change ' + state, data);
        if (state === this.state.auth) { return; }

        if (state === 'signedOut') { state = 'signIn'; }
        try {
            localStorage.setItem(AUTHENTICATOR_AUTHSTATE, state);
        } catch (e) {
            logger.debug('Failed to set the auth state into local storage', e);
        }
        this.setState({ auth: state, authData: data, error: null, showToast: false });
        if (this.props.onStateChange) { this.props.onStateChange(state, data); }
    }

    handleAuthEvent(state, event, showToast = true) {
        if (event.type === 'error') {
            const map = this.props.errorMessage || AmplifyMessageMap;
            const message = (typeof map === 'string')? map : map(event.data);
            this.setState({ error: message, showToast });
            
        }
    }

    render() {
        const { auth, authData } = this.state;
        const theme = this.props.theme || AmplifyTheme;
        const messageMap = this.props.errorMessage || AmplifyMessageMap;

        let { hideDefault, hide = [], federated } = this.props;
        if (hideDefault) {
            hide = hide.concat([
                Greetings,
                SignIn,
                ConfirmSignIn,
                RequireNewPassword,
                SignUp,
                ConfirmSignUp,
                VerifyContact,
                ForgotPassword,
                TOTPSetup
            ]);
        }
        const props_children = this.props.children || [];

        const default_children = [
            <Greetings federated={federated}/>,
            <SignIn federated={federated}/>,
            <ConfirmSignIn/>,
            <RequireNewPassword/>,
            <SignUp/>,
            <ConfirmSignUp/>,
            <VerifyContact/>,
            <ForgotPassword/>,
            <TOTPSetup/>
        ];

        const props_children_names  = React.Children.map(props_children, child => child.type.name);
        const props_children_override =  React.Children.map(props_children, child => child.props.override);
        hide = hide.filter((component) =>!props_children_names.includes(component.name));
        const hideLink = hide.filter((component) => {
            return !props_children_override.some(comp => comp === component);
        });
        
        const render_props_children = React.Children.map(props_children, (child, index) => {
            return React.cloneElement(child, {
                    key: 'aws-amplify-authenticator-props-children-' + index,
                    theme,
                    messageMap,
                    authState: auth,
                    authData,
                    onStateChange: this.handleStateChange,
                    onAuthEvent: this.handleAuthEvent,
                    hide: hide,
                    hideLink: hideLink
                });
        });
       
        const render_default_children = hideDefault ? [] : React.Children.map(default_children, (child, index) => {
                return React.cloneElement(child, {
                    key: 'aws-amplify-authenticator-default-children-' + index,
                    theme,
                    messageMap,
                    authState: auth,
                    authData,
                    onStateChange: this.handleStateChange,
                    onAuthEvent: this.handleAuthEvent,
                    hide: hide,
                    hideLink: hideLink
                });
            });

        const render_children = render_default_children.concat(render_props_children);
        const error = this.state.error;        

        return (
            <Container theme={theme}>
                {this.state.showToast && 
                    <Toast theme={theme} onClose={() => this.setState({showToast: false})}>
                        { I18n.get(error) }
                    </Toast>
                }
                {render_children}
            </Container>
        );
    }
}
