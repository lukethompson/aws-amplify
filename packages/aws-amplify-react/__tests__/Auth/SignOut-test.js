import Auth from '@aws-amplify/auth';
import SignOut from '../../src/Auth/SignOut';
import * as React from 'react';
import AmplifyTheme from '../../src/AmplifyTheme';
import AuthPiece from '../../src/Auth/AuthPiece';
import { Header, Footer, InputRow, ButtonRow } from '../../src/AmplifyUI';

const acceptedStates = [
    'signedIn'
];

const deniedStates = [
    'signIn',  
    'signedUp', 
    'signedOut',
    'forgotPassword',
    'signUp',
    'confirmSignIn',
    'confirmSignUp',
    'verifyContact'
];

describe('SignOut', () => {
    describe('normal case', () => {
        test('render correctly with authState signedIn', () => {
            const wrapper = shallow(<SignOut/>);
            for (let i = 0; i < acceptedStates.length; i += 1){
                wrapper.setProps({
                    authState: acceptedStates[i],
                    theme: 'theme'
                });
                expect(wrapper).toMatchSnapshot();
            }
        });

        test('render correctly with hide', () => {
            const wrapper = shallow(<SignOut/>);
            for (let i = 0; i < acceptedStates.length; i += 1){
                wrapper.setProps({
                    authState: acceptedStates[i],
                    theme: 'theme',
                    hide: [SignOut]
                });
                expect(wrapper).toMatchSnapshot();
            }
        });

        test('render correctly with empty hide', () => {
            const wrapper = shallow(<SignOut/>);
            for (let i = 0; i < acceptedStates.length; i += 1){
                wrapper.setProps({
                    authState: acceptedStates[i],
                    theme: 'theme',
                    hide: []
                });
                expect(wrapper).toMatchSnapshot();
            }
        });
   
    test('render correctly with other authStates', () => {
        const wrapper = shallow(<SignOut/>);
        
        for (let i = 0; i < deniedStates.length; i += 1){
            wrapper.setProps({
                authState: deniedStates[i],
                theme: 'theme'
            });

            expect(wrapper).toMatchSnapshot();
        }
    });

    describe('signOut test', () => {
        test('happy case', async () => {
            const wrapper = shallow(<SignOut/>);
            const signOut = wrapper.instance();

            const spyon = jest.spyOn(Auth, 'signOut').mockImplementationOnce(() => {
                return Promise.resolve();
            });

            await signOut.signOut();

            expect(spyon).toBeCalled();
            spyon.mockClear();
        });

        test('error case', async () => {
            const wrapper = shallow(<SignOut/>);
            const signOut = wrapper.instance();

            const spyon = jest.spyOn(Auth, 'signOut').mockImplementationOnce(() => {
                return Promise.reject('error');
            });

            await signOut.signOut();

            expect(spyon).toBeCalled();
            spyon.mockClear();
        });
    });
});
