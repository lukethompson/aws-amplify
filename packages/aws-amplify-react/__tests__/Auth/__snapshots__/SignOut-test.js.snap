// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignOut normal case render correctly with authState signedIn 1`] = `""`;

exports[`SignOut normal case render correctly with empty hide 1`] = `""`;

exports[`SignOut normal case render correctly with hide 1`] = `""`;

exports[`SignOut normal case render correctly with other authStates 1`] = `""`;

exports[`SignOut normal case render correctly with other authStates 2`] = `""`;

exports[`SignOut normal case render correctly with other authStates 3`] = `""`;

exports[`SignOut normal case render correctly with other authStates 4`] = `""`;

exports[`SignOut normal case render correctly with other authStates 5`] = `""`;

exports[`SignOut normal case render correctly with other authStates 6`] = `""`;

exports[`SignOut normal case render correctly with other authStates 7`] = `""`;

exports[`SignOut normal case render correctly with other authStates 8`] = `""`;
